Sources/SwiftFusion/Inference/ChordalInitialization.swift

```diff
@@ -0,0 +1,213 @@
+    }
+    
+    associations[anchorId.perTypeID] = orientations.store(Vector9(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0))
+    
+    for factor in g.factors(type: BetweenFactor3.self) {
+      let R = factor.difference.rot.coordinate.R
+      let R_v = Vector9(R.s00, R.s01, R.s02, R.s10, R.s11, R.s12, R.s20, R.s21, R.s22)
+      let frob_factor = FrobeniusFactorRot3(associations[factor.edges.head.perTypeID]!, associations[factor.edges.tail.head.perTypeID]!, R_v)
+      orientationGraph.store(frob_factor)
+    }
+    
+    orientationGraph.store(FrobeniusAnchorFactorRot3(associations[anchorId.perTypeID]!, Vector9(1.0, 0.0, 0.0, /*  */ 0.0, 1.0, 0.0, /*  */ 0.0, 0.0, 1.0)))
+    
+    var optimizer = GenericCGLS()
+    let linearGraph = orientationGraph.linearized(at: orientations)
+    optimizer.optimize(gfg: linearGraph, initial: &orientations)
+    
+    return normalizeRelaxedRotations(orientations, associations: associations, ids: ids)
+  }
+  
+  /// This function finds the closest Rot3 to the unconstrained 3x3 matrix with SVD.
+  /// TODO(fan): replace this with a 3x3 specialized SVD instead of this generic SVD (slow)
+  public func normalizeRelaxedRotations(
+    _ relaxedRot3: VariableAssignments,
+    associations: Dictionary<Int, TypedID<Vector9, Int>>,
+    ids: Array<TypedID<Pose3, Int>>) -> VariableAssignments {
+    var validRot3 = VariableAssignments()
+    
+    for v in ids {
+      let M_v: Vector9 = relaxedRot3[associations[v.perTypeID]!]
+      
+      let M = Matrix3(M_v.s0, M_v.s1, M_v.s2, M_v.s3, M_v.s4, M_v.s5, M_v.s6, M_v.s7, M_v.s8)
+      
+      let initRot = Rot3.ClosestTo(mat: M)
+      
+      // TODO(fan): relies on the assumption of continous and ordered allocation
+      let _ = validRot3.store(initRot)
+    }
+    return validRot3;
+  }
+  
+  /// This function computes the inital poses given the chordal initialized rotations.
+  public func computePoses(graph: FactorGraph, orientations: VariableAssignments, ids: Array<TypedID<Pose3, Int>>) -> VariableAssignments {
+    var val = VariableAssignments()
+    for v in ids {
+      let _ = val.store(Pose3(orientations[TypedID<Rot3, Int>(v.perTypeID)], Vector3(0,0,0)))
+    }
+    
+    // optimize for 1 G-N iteration
+    for _ in 0..<1 {
+      let gfg = graph.linearized(at: val)
+      var dx = val.tangentVectorZeros
+      var optimizer = GenericCGLS(precision: 1e-6, max_iteration: 500)
+      optimizer.optimize(gfg: gfg, initial: &dx)
+      val.move(along: (-1) * dx)
+    }
+    return val
+  }
+  
+  /// This function computes the chordal initialization. Normally this is what the user needs to call.
+  /// TODO(fan): This function builds upon the assumption that all variables stored are Pose3s, could fail if that is not the case.
+  public mutating func GetInitializations(graph: FactorGraph, val: VariableAssignments, ids: Array<TypedID<Pose3, Int>>) -> VariableAssignments {
```

Comment:
Document arguments. In particular, why do I need a val? I though chordal *created* a val.

Reply:
done